pkg/operator/csr/csr_approver.go

```diff
@@ -0,0 +1,272 @@
+		}
+
+		csrDecision, denyReason, err := c.csrApprover.Approve(csr, x509CSR)
+		if err != nil {
+			return c.denyCSR(ctx, csrCopy, "CSRApprovingFailed", fmt.Sprintf("there was an error during CSR approval: %v", err), syncCtx.Recorder())
+		}
+
+		switch csrDecision {
+		case CSRDenied:
+			return c.denyCSR(ctx, csrCopy, "CSRDenied", denyReason, syncCtx.Recorder())
+		case CSRApproved:
+			csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
+				certapiv1.CertificateSigningRequestCondition{
+					Type:    certapiv1.CertificateApproved,
+					Status:  corev1.ConditionTrue,
+					Reason:  "AutoApproved",
+					Message: fmt.Sprintf("Auto-approved CSR %q", csr.Name),
+				},
+			)
+
+			syncCtx.Recorder().Eventf("CSRDeny", "The CSR %q has been approved", csrCopy.Name)
+			_, err = c.csrClient.UpdateApproval(ctx, csr.Name, csrCopy, v1.UpdateOptions{})
+			return err
+		case CSRNoOpinion:
+			fallthrough
+		default:
+			return nil
+		}
+	}
+
+	// no matching CSRs were found
+	return nil
+}
+
+func (c *csrApproverController) denyCSR(ctx context.Context, csrCopy *certapiv1.CertificateSigningRequest, reason, message string, eventsRecorder events.Recorder) error {
+	csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
+		certapiv1.CertificateSigningRequestCondition{
+			Type:    certapiv1.CertificateDenied,
+			Status:  corev1.ConditionTrue,
+			Reason:  reason,
+			Message: message,
+		},
+	)
+
+	eventsRecorder.Eventf("CSRDeny", "The CSR %q has been denied", csrCopy.Name)
+	_, err := c.csrClient.UpdateApproval(ctx, csrCopy.Name, csrCopy, v1.UpdateOptions{})
+	return err
+}
+
+func getCertApprovalCondition(status *certapiv1.CertificateSigningRequestStatus) (approved bool, denied bool) {
+	for _, c := range status.Conditions {
+		if c.Type == certapiv1.CertificateApproved {
+			approved = true
+		}
+		if c.Type == certapiv1.CertificateDenied {
+			denied = true
+		}
+	}
+	return
+}
+
+type ServiceAccountApprover struct {
+	saGroups        sets.String // saGroups is the set of groups for the SA expected to have created the CSR
+	saName          string
+	expectedSubject string
+}
+
+// ServiceAccountApprover approves CSRs with a given subject issued by the provided service account
+func NewServiceAccountApprover(saNamespace, saName, expectedSubject string) *ServiceAccountApprover {
+	return &ServiceAccountApprover{
+		saName:          serviceaccount.MakeUsername(saNamespace, saName),
+		saGroups:        sets.NewString(append(serviceaccount.MakeGroupNames(saNamespace), "system:authenticated")...),
+		expectedSubject: expectedSubject,
+	}
+}
+
+func (a *ServiceAccountApprover) Approve(csrObj *certapiv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) (approvalStatus CSRApprovalDecision, denyReason string, err error) {
+	if csrObj == nil || x509CSR == nil {
+		return CSRDenied, "", fmt.Errorf("received a 'nil' CSR")
```

Comment:
as we are denying here, we should return a deny reason here as well, no?

Reply:
I can add an "Error" deny reason here